server/src/test/java/com/microsoft/java/bs/core/internal/server/BuildTargetServerIntegrationTest.java

```diff
@@ -53,55 +184,122 @@ static void beforeClass() {
     System.setProperty(Launcher.PROP_PLUGIN_DIR, pluginDir);
   }
 
+  @AfterAll
+  static void afterClass() {
+    System.clearProperty(Launcher.PROP_PLUGIN_DIR);
+  }
+
   @BeforeEach
-  void setUp() {
+  void beforeEach() throws IOException {
+    threadPool = Executors.newCachedThreadPool();
+    PipedInputStream clientIn = new PipedInputStream();
+    clientOut = new PipedOutputStream();
+    PipedInputStream serverIn = new PipedInputStream();
+    serverOut = new PipedOutputStream();
+    clientIn.connect(serverOut);
+    clientOut.connect(serverIn);
+    // server
     BuildTargetManager buildTargetManager = new BuildTargetManager();
     PreferenceManager preferenceManager = new PreferenceManager();
     GradleApiConnector connector = new GradleApiConnector(preferenceManager);
     LifecycleService lifecycleService = new LifecycleService(connector, preferenceManager);
     BuildTargetService buildTargetService = new BuildTargetService(buildTargetManager,
         connector, preferenceManager);
-    gradleBuildServer = new GradleBuildServer(lifecycleService, buildTargetService);
+    GradleBuildServer gradleBuildServer =
+        new GradleBuildServer(lifecycleService, buildTargetService);
+    org.eclipse.lsp4j.jsonrpc.Launcher<BuildClient> serverLauncher =
+        new org.eclipse.lsp4j.jsonrpc.Launcher.Builder<BuildClient>()
+            .setLocalService(gradleBuildServer)
+            .setRemoteInterface(BuildClient.class)
+            .setOutput(serverOut)
+            .setInput(serverIn)
+            .setExecutorService(threadPool)
+            .create();
+    buildTargetService.setClient(serverLauncher.getRemoteProxy());
+    // client
+    client = new TestClient();
+    clientLauncher = new org.eclipse.lsp4j.jsonrpc.Launcher.Builder<TestServer>()
+            .setLocalService(client)
+            .setRemoteInterface(TestServer.class)
+            .setInput(clientIn)
+            .setOutput(clientOut)
+            .setExecutorService(threadPool)
+            .create();
+    // start
+    clientLauncher.startListening();
+    serverLauncher.startListening();
   }
 
-  @AfterAll
-  static void afterClass() {
-    System.clearProperty(Launcher.PROP_PLUGIN_DIR);
+  @AfterEach
+  void afterEach() {
+    // closing the streams and pool shuts down the servers
+    try {
+      clientOut.close();
+      /*
+         Gradle will still fire messages for a while after the connection is closed
```

Comment:
Should we call the `.buildShutdown()` for the server? Will this help with the output stream issue mentioned here?

Reply:
I'd forgotten that existed.  I've added that in and refactored so that the streams are closed using try-with-resources.  I now can't reproduce the issue I had before.